File: entity-meta-generator-lib/src/main/java/com/loudsight/meta/EntityInstantiator.java
package com.loudsight.meta;

import java.util.Collection;
import java.util.List;
import java.util.Map;

import com.loudsight.meta.entity.EntityConstructor;
import com.loudsight.meta.entity.EntityField;
import com.loudsight.meta.entity.EntityParameter;
import com.loudsight.meta.serialization.TypeConverters;

public class EntityInstantiator {

    private static class EntityInstantiatorHolder {
        private static final EntityInstantiator INSTANCE = new EntityInstantiator();
    }
    // global access point
    public static EntityInstantiator getInstance() {
        return EntityInstantiator.EntityInstantiatorHolder.INSTANCE;
    }
    private EntityInstantiator() {
    }

    private <T> Object[] getParameters(EntityConstructor constructor,
                                  Map<String, EntityField<T, ?>> fieldsMap,
                                  Map<String, ?> entityMap) {
        return constructor.getEntityParameters()
                .stream()
            .map(it -> {
                var fieldName = it.name();
//            var type = it.parameterType.qualifiedName!!
                var fieldValue = entityMap.get(fieldName);

                if (fieldValue == null) {
                    return null;
                }
                var field = fieldsMap.get(fieldName);
                if (field == null) {
                    return fieldValue;
                }

                Object convertedValue;
                if (field.isEnum() && fieldValue instanceof String enumName) {
                    convertedValue = convertEnumValue(field.typeClass(), enumName);
                } else {
                    convertedValue = TypeConverters.getInstance().convert(fieldValue, field.typeClass());
                }

                return convertedValue;
            })
            .toArray();
    }

    @SuppressWarnings("unchecked")
    public <T> T invoke(Collection<EntityConstructor> constructors,
                        Map<String, EntityField<T, ?>> fieldsMap,
                        Map<String, ?> paramsMap) {
        int size;
        if (paramsMap.containsKey("__className__")) {
            size = paramsMap.size() - 1;
        } else {
            size = paramsMap.size();
        }

        // If paramsMap is empty, try no-arg constructor directly
        if (size == 0) {
            var c = constructors
                    .stream()
                    .filter(it -> it.getEntityParameters().isEmpty())
                    .toList();

            if (!c.isEmpty()) {
                return c.get(0).<T>newInstance();
            }
            throw new IllegalStateException("Cannot invoke any constructors with " + paramsMap);
        }

        // Try exact match first
        var c = findConstructors(constructors, paramsMap, size, true);
        if (!c.isEmpty()) {
            return instantiateWithConstructor(c.get(0), fieldsMap, paramsMap);
        }

        // Fallback: try constructors matching subset of parameters
        c = findConstructors(constructors, paramsMap, size, false);
        if (!c.isEmpty()) {
            return instantiateWithConstructor(c.get(0), fieldsMap, paramsMap);
        }

        // Try no-arg constructor
        c = constructors
                .stream()
                .filter(it -> it.getEntityParameters().isEmpty())
                .toList();

        if (c.isEmpty()) {
            throw new IllegalStateException("Cannot invoke any constructors with " + paramsMap);
        }

        var entity = c.get(0).<T>newInstance();

        paramsMap
                .entrySet()
                .stream()
                .filter(e -> fieldsMap.containsKey(e.getKey()))
                .forEach(e -> {
                    var fieldName = e.getKey();
                    var fieldValue = e.getValue();
                    var field = fieldsMap.get(fieldName);
                    Object convertedValue;
                    if (field.isEnum() && fieldValue instanceof String enumName) {
                        convertedValue = convertEnumValue(field.typeClass(), enumName);
                    } else {
                        convertedValue = TypeConverters.getInstance().convert(fieldValue, field.typeClass());
                    }
                    field.set(entity, convertedValue);
                });
        return entity;
    }

    private <T> void setExtraFields(T entity, EntityConstructor constructor,
                                     Map<String, EntityField<T, ?>> fieldsMap,
                                     Map<String, ?> paramsMap) {
        paramsMap.entrySet().stream()
                .filter(e -> fieldsMap.containsKey(e.getKey()) && fieldsMap.get(e.getKey()).setter() != null)
                .filter(e -> !constructor.getEntityParameters().stream().map(EntityParameter::name).toList().contains(e.getKey()))
                .forEach(e -> {
                    var field = fieldsMap.get(e.getKey());
                    Object convertedValue;
                    if (field.isEnum() && e.getValue() instanceof String enumName) {
                        convertedValue = convertEnumValue(field.typeClass(), enumName);
                    } else {
                        convertedValue = TypeConverters.getInstance().convert(e.getValue(), field.typeClass());
                    }
                    field.set(entity, convertedValue);
                });
    }

    /**
     * Converts a wire-degraded enum name String back to a real enum instance. Throws a clear,
     * immediate exception if the enum type isn't resolvable on this JVM, rather than letting a
     * NullPointerException surface later and further from the actual cause (or worse, silently
     * dropping the field - setExtraFields only sets fields it successfully converts).
     */
    private Object convertEnumValue(Class<?> enumType, String enumName) {
        var meta = MetaRepository.getInstance().getMeta(enumType);
        if (meta == null) {
            throw new IllegalStateException("Cannot resolve Meta for enum type " + enumType.getName()
                    + " to convert value \"" + enumName + "\" - the enum type is not on this JVM's classpath");
        }
        return meta.newInstance(Map.of("name", enumName));
    }

    private List<EntityConstructor> findConstructors(Collection<EntityConstructor> constructors,
                                                      Map<String, ?> paramsMap, int size, boolean exactMatch) {
        return constructors
                .stream()
                .filter(it -> !it.getEntityParameters().isEmpty() && paramsMap.keySet().containsAll(it.getEntityParameters().stream().map(EntityParameter::name).toList()))
                .filter(it -> exactMatch ? it.getEntityParameters().size() >= size : it.getEntityParameters().size() < size)
                .sorted((a, b) -> Integer.compare(b.getEntityParameters().size(), a.getEntityParameters().size()))
                .toList();
    }

    @SuppressWarnings("unchecked")
    private <T> T instantiateWithConstructor(EntityConstructor constructor,
                                              Map<String, EntityField<T, ?>> fieldsMap,
                                              Map<String, ?> paramsMap) {
        var parameters = getParameters(constructor, fieldsMap, paramsMap);
        var entity = constructor.<T>newInstance(parameters);
        setExtraFields(entity, constructor, fieldsMap, paramsMap);
        return entity;
    }
}

File: entity-meta-generator-lib/src/main/kotlin/com/loudsight/meta/EntityInstantiator.kt
package com.loudsight.meta

import com.loudsight.meta.entity.EntityConstructor
import com.loudsight.meta.entity.EntityField
import com.loudsight.meta.serialization.TypeConverters
import com.loudsight.useful.helper.ClassHelper
import kotlin.reflect.KClass

object EntityInstantiator {

    private fun <T> getParameters(constructor: EntityConstructor,
                                  fieldsMap: Map<String, EntityField<T, *>>,
                                  entityMap: Map<String, Any>): Array<Any?> {
        return constructor.entityParameters
            .map {
                val fieldName = it.name
//            var type = it.parameterType.qualifiedName!!
                val fieldValue = entityMap[fieldName]

                if (fieldValue == null) {
                    return@map null
                }
                val field = fieldsMap[fieldName]
                if (field == null) {
                    return@map fieldValue
                }

                val convertedValue = if (field.isEnum && fieldValue is String) {
                    val meta = MetaRepository.getMeta(field.typeClass)!!
                    meta.newInstance(mapOf(kotlin.Pair("name", fieldValue)))
                } else {
                    TypeConverters.convert(fieldValue, field.typeClass.kotlin as KClass<Any>)!!
                }
//
                return@map convertedValue
//            } else {
//                type = type.replace("<out ", "<")
////                val e = EntityTransforms.getEntityType(it.parameterType.java, value)
//                if (type.startsWith("kotlin.Array<")) {
//                    value = helper.eval(type.replace("kotlin.Array", "arrayOf") + "()")
//                    return@map value
//                } else if (type.startsWith("kotlin.collections.List<")) {
//                    var expr = type.replace("kotlin.collections.List", "arrayListOf") + "()"
//                    expr = expr.replace("<T", "<*")
//                    value = helper.eval(expr)
//                    return@map value
//                } else if (type.startsWith("kotlin.collections.Collection<")) {
//                    var expr = type.replace("kotlin.collections.Collection", "arrayListOf") + "()"
//                    expr = expr.replace("<T", "<*")
//                    value = helper.eval(expr)
//                    return@map value
//                }
//                return@map value
//            }
            }
            .toTypedArray()
    }

    fun <T> invoke(constructors: Collection<EntityConstructor>,
                   fieldsMap: Map<String, EntityField<T, *>>,
                   paramsMap: Map<String, Any>): T {
        val size =
            if (paramsMap.containsKey("__className__")) {
                paramsMap.size - 1
            } else {
                paramsMap.size
            }

        val c = constructors
            .asSequence()
            .filter { it.entityParameters.isNotEmpty() && paramsMap.keys.containsAll(it.entityParameters.map { k -> k.name }.toList()) }
            .filter { it.entityParameters.size >= size }
            .sortedByDescending { it.entityParameters.size }
            .toList()

        c.forEach {
                val parameters = getParameters(it, fieldsMap, paramsMap)
                return@invoke it.newInstance(*parameters)
            }

        constructors
            .asSequence()
            .filter { it.entityParameters.isEmpty() }
            .first {
                val entity = it.newInstance<T>()

                paramsMap
                    .filter { (fieldName, _) -> fieldsMap.containsKey(fieldName) }
                    .forEach { (fieldName, fieldValue) ->
                    val field = fieldsMap[fieldName]!!
                    val convertedValue = if (field.isEnum && fieldValue is String) {
                        val meta = MetaRepository.getMeta(field.typeClass)!!
                        meta.newInstance(mapOf(kotlin.Pair("name", fieldValue)))
                    } else {
                        TypeConverters.convert(fieldValue, field.typeClass.kotlin )!!
                    }
                    field[entity] = convertedValue
                }
                return@invoke entity
            }

        throw IllegalStateException("Cannot invoke any constructors with $paramsMap")
    }

    fun x() {

//                val parameters = meta.constructors
//                    .first { it.entityParameters.isNotEmpty() }
//                    .entityParameters.map { parameter ->
//                    val name = parameter.name
//                    var entity = entityMap[name]
//
//                    if (entity is Map<*, *>) {
//                        entity = fromXXX(Cast.uncheckedCast(entity))
//                    } else if (entity is List<*>) {
//                        entity = (entity as List<*>).map {
//                            if (it is Map<*, *>) {
//                                fromXXX(Cast.uncheckedCast(it))
//                            } else {
//                                it
//                            }
//                        }.toList()
//                    }
//                    val value = entity
//                    if (value == null) {
//                        parameter.nullValue
//                    } else if (KClass::class.java.isAssignableFrom(parameter.parameterType.java)) {
//                        if (value is String) {
//                            Class.forName(value).kotlin
//                        } else {
//                            null
//                        }
//                    } else if (parameter.parameterType.java.isAssignableFrom(value.javaClass)) {
//                        value
//                    } else {
//                        null
//                    }

    }

//    private fun constructInstanceFromValue(value: Any?): Any? {
//        if (value is List<*>) {
//            return value.map { constructInstanceFromValue(it) }.toList()
//        } else if (value is Map<*, *> && value.containsKey("__className__")) {
//            val meta = helper.getMeta<Any>(JvmClassHelper.classForName<Any>(value["__className__"] as String))
//
//            if (meta.isPresent) {
//                return meta.get().newInstance(ClassHelper.Companion.uncheckedCast(value))
//            }
//            // value = metaRepository.fromMap(value)
//        }
//        return value
//    }
}
